### PR TITLE
#2419. Try parse SemVerInfo from .nupkg file name for localFolder nuget source

### LINKED ISSUE #2419: Paket can pick a wrong package if source is local folder
### Description

When source set to folder (local or network share) paket can pick a wrong package when perform "install" command

### Repro steps

[repro.zip](https://github.com/fsprojects/Paket/files/1072198/repro.zip)

Test setup In attached archive:
testPackage folder contains just a paket.template file for creating a test library
testSolution trying to consume created library (paket dependencies entry: `nuget testlib 1.2.3`)

Repro steps (in a attached setup):
1. go to `testPackage` folder
2. call `paket pack output ..\nuget version 1.2.3-alpha`
3. go to `testSolution` folder
4. call `paket install`

### Expected behavior

"Can't find package" error

### Actual behavior

paket create a paket.lock file with line `testlib (1.2.3)` and copy `testlib.1.2.3-alpha.nupkg` to packages folder (and you can't see it unless browse a packages folder)

### Known workarounds

using nuget server instead of folder
using something like `>= 1.2.3 <= 1.2.3` version constraint

### Other cases

If folder contains two versions of package: 0.1.2.0 and 1.2.0.0 and dependency file line like `nuget <...> 1.2` than paket pick package with version 0.1.2.0.

Repro steps (in a attached setup):
1. go to `testPackage`
2. call `paket pack output ..\nuget version 0.1.2.0`
3. call `paket pack output ..\nuget version 1.2.0.0`
4. go to `testSolution` folder
5. add `nuget testlib 1.2` to paket.dependencies file
6. call `paket install`

Expected: package `testlib 1.2.0.0` in packages folder
Actual: package `testlib 0.1.2.0` in packages folder
Workarounds: write full version in `paket.dependencies` file (1.2.0)

---------------------

I found code in `C:\sources\github\Paket\src\Paket.Core\Dependencies\NuGetLocal.fs` file, function `findLocalPackage`:

```fsharp
    let v3 =
        Directory.EnumerateFiles(directory,"*.nupkg",SearchOption.AllDirectories)
        |> Seq.map (fun x -> FileInfo(x))
        |> Seq.filter (fun fi -> String.containsIgnoreCase (packageName.CompareString)  fi.Name)
        |> Seq.filter (fun fi -> fi.Name.Contains(normalizedVersion) || fi.Name.Contains(version.ToString()))
        |> Seq.tryHead
```
Is there valid cases with arbitrary .nupkg file name format? 
Maybe paket should just stick to `<id>.<version>.nupgk` name format?

### PR DESCRIPTION
**Pull request by lexarchik**
fixes #2419

### CODE REVIEW

**Comment by forki** [comment]
wow this looks very good!

File: src/Paket.Core/Dependencies/NuGetLocal.fs [hunk 1]
```diff
@@ -41,16 +52,18 @@ let findLocalPackage directory (packageName:PackageName) (version:SemVerInfo) =
     let v2 = FileInfo(Path.Combine(directory, sprintf "%O.%s.nupkg" packageName normalizedVersion))
     if v2.Exists then v2 else
 
+    let condition x = 
+        match parsePackageInfoFromFileName x with
+        | Some (name, ver) -> packageName.Equals name && version.Equals ver
```
Comment (vasily-kirichenko) [inline src/Paket.Core/Dependencies/NuGetLocal.fs]: Why not `| Some (name, ver) -> packageName = name && version = ver`?

File: tests/Paket.Tests/NuGetLocal/NuGetLocalSpecs.fs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+[<Test>]
+let ``package file name split to id and version correctly``() =
+    [ 
+        ("package.1.2.3.nupkg", Some (PackageName "package", SemVer.Parse "1.2.3"));
+        ("package.1.2.3-alpha2.nupkg", Some (PackageName "package", SemVer.Parse "1.2.3-alpha2"));
+        ("package.name.with.dots.1.2.3-alpha2.nupkg", Some (PackageName "package.name.with.dots", SemVer.Parse "1.2.3-alpha2")); 
+        ("package.1.2.3.nupkg.back", None);
+        ("package.name.without.version.nupkg", None);
+    ]
```
Comment (vasily-kirichenko) [inline tests/Paket.Tests/NuGetLocal/NuGetLocalSpecs.fs]:
```
     [ 
         "package.1.2.3.nupkg", Some (PackageName "package", SemVer.Parse "1.2.3")
         "package.1.2.3-alpha2.nupkg", Some (PackageName "package", SemVer.Parse "1.2.3-alpha2")
         "package.name.with.dots.1.2.3-alpha2.nupkg", Some (PackageName "package.name.with.dots", SemVer.Parse "1.2.3-alpha2") 
         "package.1.2.3.nupkg.back", None
         "package.name.without.version.nupkg", None
     ]
```

File: tests/Paket.Tests/NuGetLocal/NuGetLocalSpecs.fs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    |> List.iter 
+        (fun (input, expectedResult) ->
+            let actualResult = parsePackageInfoFromFileName input 
+            shouldEqual expectedResult actualResult)
+
+[<Test>]
+let ``package file picked up correctly``() = 
+    [   // folder * inputPackageId * inputPackageVersion * expectedFileName
+        "./NuGetLocal/case1", PackageName "package.name", SemVer.Parse "0.1.0-alpha2", "package.name.0.1.0-alpha2.nupkg";
```
Comment (vasily-kirichenko) [inline tests/Paket.Tests/NuGetLocal/NuGetLocalSpecs.fs]: Unnecessary semicolons.

**Comment by forki** [comment]
travis failures seem unrelated.
I guess the timeouts are because NuGet forbi to use their fastest API. ;-)

File: tests/Paket.Tests/NuGetLocal/NuGetLocalSpecs.fs (hunk 3 above)
Comment (lexarchik) [inline tests/Paket.Tests/NuGetLocal/NuGetLocalSpecs.fs]: thank you, fix all three places

**Comment by matthid** [comment]
For the travis failure we will soon have https://github.com/fsprojects/Paket/pull/2423 
Interestingly it seems that NuGet is not to blame but mono (who would have thought from the error message)...
For now we can just hammer the repeat build button until we get a fixed mono version :/

**Pull request merged**